### DISCUSSION TITLE
Upgrade lists available upgrade, but says none available for installation

### DISCUSSION THREAD

**Discussion by gerhard4** [microsoft/winget-cli#3463] [Q&A] [answered]
```
❯ winget upgrade | findstr pgAdmin
pgAdmin 4 version 6.19            PostgreSQL.pgAdmin      6.19       7.4        winget


C:\cmd took 2s
❯ winget upgrade postgresql.pgadmin
No available upgrade found.
No newer package versions are available from the configured sources.
```

AIUI, the output of `winget upgrade` shows a newer package version to be available (7.4). But then the actual attempt to upgrade says "No newer package versions are available". 

How do I reconcile the two?

**Comment by Trenly** [comment]
What is the output of `winget list pgAdmin`

**Reply by gerhard4** [reply]
Thanks for responding. This is interesting:
```
❯ winget list pgadmin
Name                   Id                 Version Available Source
------------------------------------------------------------------
pgAdmin 4 version 6.19 PostgreSQL.pgAdmin 6.19    7.4       winget
pgAdmin 4 version 7.4  PostgreSQL.pgAdmin 7.4               winget
```
What does this mean? Did maybe winget install a second copy of pgAdmin when I ran the upgrade, instead of upgrading the existing installation?

**Reply by Trenly** [reply, marked as answer]
Some applications install side-by-side, and Winget doesn’t yet have a good way to reason about this behavior. Because you have an older version installed, it detects that the older version can be upgraded, but when going to perform the upgrade it detects the newer version is already installed and provides the message that no newer versions are available

**Reply by gerhard4** [reply]
Thank you! I fixed this by uninstalling the older version through the Settings tool.

**Comment by carehart** [comment]
While @Trenly had a great answer for @gerhard4's situation, I'll offer here an example where that's not the solution. Perhaps others will find this and benefit hearing if there's a different solution or tips.

In my case, it's an implementation of a VC++ Redistributable, `Microsoft.VCRedist.2015+.x86`.  While `upgrade`  shows there is an available update (from `winget` as `source`):
```
Microsoft Visual C++ 2015-2022 Redistributable (x64) … Microsoft.VCRedist.2015+.x86 14.34.31938.0 14.38.33130.0  winget`
```

running `upgrade Microsoft.VCRedist.2015+.x86` gets the error @gerhard4 noted, yet doing a `list` on it shows indeed only the one package:
```
Name                                                               Id                           Version       Source
---------------------------------------------------------------------------------------------------------------------
Microsoft Visual C++ 2015-2022 Redistributable (x86) - 14.38.33130 Microsoft.VCRedist.2015+.x86 14.38.33130.0 winget
```

Anyone have thoughts on understanding or resolving that? :-)

**Reply by Trenly** [reply]
What specific command did you use for running the list? Have you tried `winget list "2015-2022 Redistributable"`?

**Reply by carehart** [reply]
Ah, I just noticed something as I looked more closely at the `upgrade` results, it shows x64 for the name, bit x86 in the id. That seems (I suppose) some sort of mistake in winget's processing of the repo (or in the repo's package details). Before I or others may dig into that, I just wanted get this word out before anyone might see that last note but not yet notice that distinction.

More specifically, before even noticing that, I observed the 14.34 vs 14.38 versions (in the `upgrade` output,) and I wondered if perhaps those really were different "versions" of the app. So I thought to look at the Windows "add or remove programs" feature, to see if I might well have two versions (similar to what @Trenly originally proposed). I searched for `c++ 2015` and indeed it showed:
![image](https://github.com/microsoft/winget-cli/assets/389746/7eced01d-d183-4b73-bf29-6943103e3466)

And that's when I noticed first that **one was x86 and one x64**. More than that, they each had the versions shown in the `upgrade` output, with the x64 app having the lower value (starting with 14.34) and the x86 app having the higher value (14.38). Interesting.

So again, this at least explains some of what this was showing, but it still begs the question of what's amiss, in winget or the repo. For now, I will leave that for others to explore. I need to get on to other work, but I thought it may be useful for readers of this thread, who may find it in the future with their own curious variants of this error. 

(It does seem that another opp for improvement would be for the error message to be more clear. Both @gerhard4 and I were thrown off by it saying simply, "No newer package versions are available from the configured sources", which conflicted with the output of the `upgrade` command.)

**Reply by carehart** [reply]
@Trenly, I was writing my last reply while your response had come in--and the web UI did not make it clear to me before submitting. To answer your question, I had just used the id shown in the `upgrade` output, so:
```winget list Microsoft.VCRedist.2015+.x86```

**Reply by Trenly** [reply]
> Ah, I just noticed something as I looked more closely at the upgrade results, it shows x64 for the name, bit x86 in the id. That seems (I suppose) some sort of mistake in winget's processing of the repo (or in the repo's package details). Before I or others may dig into that, I just wanted get this word out before anyone might see that last note but not yet notice that distinction.

This is a known issue. Take a look at [this search](https://github.com/microsoft/winget-cli/issues?q=is%3Aopen+is%3Aissue+label%3AArea-Matching+VCRedist+) and the related issues in the https://github.com/microsoft/winget-cli/labels/Area-Matching label.

The matching issue is 100% what is causing your problems with the upgrade command

**Reply by carehart** [reply]
Also, I find someone else raising an issue like mine in [another discussion here](https://github.com/microsoft/winget-pkgs/issues/55576) (in the winget-pkgs repo), where @OfficialEsco offers an explanation from Mar 2022 that seems still to be the issue today.

**Reply by carehart** [reply]
Again, we're neck and neck to the finish line. :-) Thanks for the broader search which found the additional ones. I guess for now, "it is what it is". Again, I'm glad to have shared the experience here, as my original search for the error message led me to this as a first result in Google. I hope what I added may help others, and then what you added may help those who may want to dig in still further. Thanks again.
